src/ripple/basics/algorithm.h

```diff
@@ -0,0 +1,98 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2019 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_ALGORITHM_H_INCLUDED
+#define RIPPLE_ALGORITHM_H_INCLUDED
+
+#include <utility>
+
+namespace ripple {
+
+// For each pair of elements {i, j} in the intersection of the sorted sequences
+// [first1, last1) and [first2, last2), perform action(i, j).
+template <class InputIter1, class InputIter2, class Action, class Comp>
+void
+generalized_set_intersection(InputIter1 first1, InputIter1 last1,
+                             InputIter2 first2, InputIter2 last2,
+                             Action action, Comp comp)
+{
+    while (first1 != last1 && first2 != last2)
+    {
+        
+        if (comp(*first1, *first2))       // if *first1 < *first2
+            ++first1;                     //     then reduce first range
+        else
+        {
+            if (!comp(*first2, *first1))  // if *first1 == *first2
+            {                             //     then this is an intersection
+                action(*first1, *first2); //     do the action
+                ++first1;                 //     reduce first range
+            }
+            ++first2;        // Reduce second range because *first2 <= *first1
+        }
+    }
+}
+
+// Eliminates all the elements i in the range [first1, last1) which
+// are equivalent to some value in [first2, last2) or for which
+// pred(i) returns true.
+template <class FwdIter1, class InputIter2, class Pred, class Comp>
+FwdIter1
+remove_if_intersect_or_match(FwdIter1 first1, FwdIter1 last1,
+                             InputIter2 first2, InputIter2 last2,
+                             Pred pred, Comp comp)
+{
```

Comment:
As far as I can tell the algorithm does exactly what you want.

Unfortunately, due to my limited brain capacity, I had trouble reading it again.  I think my biggest problem was with the iterator names.  The range `[first1, last1)` is being removed from, but `[first2, last2)` is unmodified.  So I renamed them `rmFrom_first/last` and `cmpTo_first/last`.  That helped.

I also found things easier to think about if I left `rmFrom_first` unmodified.  So I copied `rmFrom_first` to a new iterator named `preserved`.

Also, just to help me think about stuff, I made as many of the passed in iterators `const` as I could.

I think that's most of what I changed.  You can see the results here if you're curious: https://github.com/scottschurr/rippled/commit/e4bf7f4a58be3048ed6ca9747e58f1ca07a3833a